flink-connectors/flink-connector-mongodb/src/main/java/org/apache/flink/mongodb/streaming/sink/MongoBulkWriter.java

```diff
@@ -0,0 +1,272 @@
+package org.apache.flink.mongodb.streaming.sink;
+
+import org.apache.flink.api.connector.sink.SinkWriter;
+import org.apache.flink.mongodb.streaming.serde.DocumentSerializer;
+import org.apache.flink.util.concurrent.ExecutorThreadFactory;
+
+import com.mongodb.MongoException;
+import com.mongodb.bulk.BulkWriteResult;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.model.BulkWriteOptions;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.UpdateOneModel;
+import com.mongodb.client.model.UpdateOptions;
+import com.mongodb.client.model.WriteModel;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+public class MongoBulkWriter<IN> implements SinkWriter<IN, DocumentBulk, DocumentBulk> {
```

Comment:
This looks like you can replace it with [AsyncSinkWriter](https://github.com/apache/flink/blob/master/flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java#L51-L51).

Reply:
@AHeise Your suggestion is great, I have already started to implement it, and I expect to finish this by the end of the week, so I'll have to ask you to look at it then.

Reply:
@AHeise I looked at the test case for ArrayListAsyncSink. I have implemented MongoAsyncSink, and I hope you can help me to see if I have implemented it correctly.